src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2304,31 +2304,53 @@ void MarkGenericParameter (GenericParameter parameter)
+				if (!IgnoreScope (@base.DeclaringType.Scope) && !IsMethodNeededByTypeDueToPreservedScope (@base))
 					continue;
 
-				// If the type is marked, we need to keep overrides of abstract members defined in assemblies
-				// that are copied.  However, if the base method is virtual, then we don't need to keep the override
-				// until the type could be instantiated
-				if (!@base.IsAbstract)
+				// ----------------------------------------------------------------------------------
+				// Base type: | base meth  | base meth | base type is     | Method scope w | method is
+				// base/iface | inst/stat  | virt/abst | iface w/ static  | UnusedIFaceOpt | needed
+				// ----------------------------------------------------------------------------------
+				// base       | instance   | virtual   | n/a              | n/a            | no
+				// base       | instance   | abstract  | n/a              | n/a            | yes
+				// base       | static     | *         | n/a              | n/a            | yes
+				// iface      | instance   | virtual   | *                | *              | no
+				// iface      | instance   | abstract  | yes              | yes            | yes
+				// iface      | instance   | *         | no               | yes            | no
+				// iface      | static     | *         | always yes       | yes            | yes
+				// iface      | *          | *         | *                | no             | yes
+
+
+				// base/iface | inst/stat  | virt/abst | iface has static | UnusedIFaceOpt | needed
+				// --------------------------------------------------------------------------------
+				// base       | instance   | virtual   | n/a              | n/a            | no
+				// Instance methods of base types only need implmentations if the base is abstract
+				// A virtual instance method with a default impl on the base doesn't need to be kept
+				if (!@base.DeclaringType.IsInterface && !@base.IsStatic && @base.IsVirtual && !@base.IsAbstract)
+					continue;
+				// base/iface | inst/stat  | virt/abst | iface has static | UnusedIFaceOpt | needed
+				// --------------------------------------------------------------------------------
+				// iface      | instance   | virtual   | *                | yes            | no
+				// Uninstantiated types will not need iface instance methods unless they are abstract
+				// Abstract iface methods must be implemented for valid IL, but not virtuals
+				if (@base.DeclaringType.IsInterface && !@base.IsStatic && @base.IsVirtual && !@base.IsAbstract)
+					continue;
+				// base/iface | inst/stat  | virt/abst | iface has static | UnusedIFaceOpt | needed
+				// --------------------------------------------------------------------------------
+				// iface      | instance   | *         | no               | yes            | no
+				// Instance methods of interfaces without static methods don't need marking
+				// It is still very possible that the interface could be removed
```

Comment:
This is unnecessarily conservative. For example:

```C#
interface IFace   // In preserved scope
{
    void Method();  // Abstract instance
    static void IfNotNull(IFace f) { if (f != null) f.Method(); }
}

class MyType : IFace
{
     void Method() { }
}

void Test()
{
    typeof(MyType); // The only use of the type
}
```

I think in this case it's OK to remove the implementation of `IFace` on `MyType` and thus remove `MyType.Method`.

I think the interesting case is if an interface has a static virtual/abstract method, then we can't really remove implementation of such interface (in theory we could, but we would have to start tracking generic instantiations and constraints, which currently we don't).

Might be worth adding a helper like `CanRemoveInterfaceImplementationForUninstantiatedType(TypeDefinition iface)` which would return true for interfaces for which we could consider removing their implementation (if the type is not instantiated). It would return false if the interface has static virtual/abstract method I think.

The above example with non-virtual static on an interface: I think we'll need to figure out if it's OK to remove that (I could be wrong above). Basically what happens if I do this:

```C#
interface IFace
{
    static void Hello () { Console.WriteLine("Hello"); }
}

class MyType : IFace
{
}

class HelloThere<T> where T : IFace
{
    public static void Greet()
    {
        T.Hello();
    }
}

public static void Test()
{
    HelloThere<MyType>.Greet();
}
```

It seems this doesn't compile, so if this is prohibited then the above holds true.

Reply:
Actually I just found out that if I change it like this:

```C#
interface IFace
{
    static void Hello () { Console.WriteLine("Hello"); }
}

class MyType : IFace
{
}

class HelloThere<T> where T : IFace
{
    public static void Greet()
    {
        // Removed the call here
    }
}

public static void Test()
{
    HelloThere<MyType>.Greet();
}
```

It now compiles, but if I remove IFace implementation from MyType it doesn't compile anymore since MyType now doesn't fulfill the generic constraint. I wonder if we have issue in linker around this at all.

Reply:
I played with this some more... linker is clever - if it sees interface in a generic constraint then such interface implementation is always preserved on all types (regardless of instantiation).
It's called "relevant to variant casting" (honestly no idea why, but that's what it does).

Reply:
So I think my above statement holds true - presence of a simple non-virtual static method on an interface doesn't mean we can't remove the interface implementation from a type.
Given the above about "relevant to variant casting" - we will never remove static interface implementations (since we would see them in generic constraints). Which fits what we discussed.

Reply:
In a way we may not need any checks for the static interface cases - due to the fact that linker will keep implementions of ifaces which are used for generic instantiation (and that is the only way static interfaces can be used really - other than reflection). We could keep the code here - as it doesn't hurt, but I would honestly expect it to work even without it (the iface implementation should be kept and since the iface is fully preserved I would expect that all implementation methods on the class are also kept - since they need to be).